vault/logical_system.go

```diff
@@ -2926,6 +2928,107 @@ func (b *SystemBackend) pathInternalUIResultantACL(ctx context.Context, req *log
 	return resp, nil
 }
 
+func (b *SystemBackend) pathInternalOpenAPI(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+
+	// Limit output to authorized paths
+	resp, err := b.pathInternalUIMountsRead(ctx, req, d)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up target document and convert to map[string]interface{} which is what will
+	// be received from plugin backends.
+	doc := openapi.NewDocument()
+
+	procMountGroup := func(group, mountPrefix string) error {
+		for mount := range resp.Data[group].(map[string]interface{}) {
+			backend := b.Core.router.MatchingBackend(ctx, mountPrefix+mount)
+
+			if backend == nil {
+				continue
+			}
+
+			req := &logical.Request{
+				Operation: logical.HelpOperation,
+			}
+
+			resp, err := backend.HandleRequest(ctx, req)
+			if err != nil {
+				return err
+			}
+
+			var backendDoc *openapi.Document
+
+			// Normalize response type, which will be different if received
+			// from and external plugin.
+			switch v := resp.Data["openapi"].(type) {
+			case *openapi.Document:
+				backendDoc = v
+			case map[string]interface{}:
+				backendDoc = new(openapi.Document)
+				if err := mapstructure.Decode(v, backendDoc); err != nil {
+					return err
+				}
+			default:
+				continue
+			}
+
+			// Prepare to add tags to default builtins that are
+			// type "unknown" and won't already be tagged.
+			var tag string
+			switch mountPrefix + mount {
+			case "cubbyhole/", "secret/":
+				tag = "secrets"
+			case "sys/":
+				tag = "system"
+			case "auth/token/":
+				tag = "auth"
+			}
+
+			// Merge backend paths with existing document
+			for path, obj := range backendDoc.Paths {
+				path := strings.TrimPrefix(path, "/")
+
+				// Add tags to all of the operations if necessary
+				if tag != "" {
+					for _, op := range []*openapi.Operation{obj.Get, obj.Post, obj.Delete} {
+						if op != nil && len(op.Tags) == 0 {
+							op.Tags = []string{tag}
```

Comment:
Should this append instead of overwrite?

Reply:
This is only hit if `len(op.Tags)==0`.  If the normal processing added any tags, don't change them here. The main OpenAPI processor, not this handler, is where the tags _should_ be added, so anything done there should be authoritative.